src/Components/Web/src/HtmlRendering/StaticHtmlRenderer.HtmlWriting.cs

```diff
@@ -323,7 +338,22 @@ private static bool TryFindEnclosingElementFrame(ArrayRange<RenderTreeFrame> fra
             }
         }
 
+        EmitFormActionIfNotExplicit(output, isForm, hasExplicitActionValue);
+
         return position + maxElements;
+
+        void EmitFormActionIfNotExplicit(TextWriter output, bool isForm, bool hasExplicitActionValue)
+        {
+            if (isForm && !hasExplicitActionValue)
+            {
+                output.Write(' ');
+                output.Write("action");
+                output.Write('=');
+                output.Write('\"');
+                _htmlEncoder.Encode(output, _navigationManager.Uri);
```

Comment:
Do we know this always matches the client's view of the URL? Just want to be sure that in URL-rewriting scenarios this doesn't generate a wrong URL. In particular in cases where people are using K8s or similar and have some HTTP ingress that forwards https://mysite.com/abc to http://internal:1234/abc

Reply:
If we know `_navigationManager.Uri` is always a relative URL then that's likely good enough, but ideally we'd do the same as whatever MVC/Razor Pages do for setting a form `action`.

Reply:
They have to account for `X-Forwarded-For` or whatever otherwise it would be broken, wouldn't it?

I don't think we can safely emit absolute URLs here, at least not ones that just come naively from `HttpContext.Request`.

Reply:
> It doesn't solve more complex URL rewriting which is why I wondered what MVC/Razor Pages do - maybe they just don't try to solve that, in which case I'd be fine with limiting it to just returning root-relative URLs.

We don't need to solve this problem. Apps need to have correctly setup their forwarding headers if behind a proxy. That's what routing relies on too for link generation. Whatever is on HttpContext is expected to be externally reachable

Reply:
OK good